bvm/ballerina-observe/src/main/java/org/ballerinalang/observe/trace/OpenTracerBallerinaWrapper.java

```diff
@@ -0,0 +1,255 @@
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*
+*/
+
+package org.ballerinalang.observe.trace;
+
+import io.opentracing.References;
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.propagation.Format;
+import org.ballerinalang.observe.trace.config.ConfigLoader;
+import org.ballerinalang.observe.trace.config.OpenTracingConfig;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * This class wraps opentracing apis and exposes native functions to use within ballerina.
+ */
+public class OpenTracerBallerinaWrapper {
+
+    static final String ROOT_CONTEXT = "root_context";
+    private static final String DEFAULT_TRACER = "default";
+    private static OpenTracerBallerinaWrapper instance = new OpenTracerBallerinaWrapper();
+    private TracersStore tracerStore;
+    private SpanStore spanStore;
+    private boolean enabled = false;
+
+    public OpenTracerBallerinaWrapper() {
+        OpenTracingConfig openTracingConfig = ConfigLoader.load();
+        if (openTracingConfig != null) {
+            tracerStore = new TracersStore(openTracingConfig);
+            enabled = true;
+        }
+        spanStore = new SpanStore();
+    }
+
+    public static OpenTracerBallerinaWrapper getInstance() {
+        return instance;
+    }
+
+    /**
+     * Method to create an entry in span store by extracting a spanContext from a Map carrier.
+     *
+     * @param spanHeaders map of headers used to extract a spanContext
+     * @return the Id of the span context
+     */
+    public String extract(Map<String, String> spanHeaders) {
+        String parentSpanId = null;
+        if (enabled) {
+            Map<String, SpanContext> spanContextMap = new HashMap<>();
+            Map<String, Tracer> tracers = tracerStore.getTracers(DEFAULT_TRACER);
+            boolean hasParent = true;
+
+            for (Map.Entry<String, Tracer> tracerEntry : tracers.entrySet()) {
+                spanContextMap.put(tracerEntry.getKey(), tracerEntry.getValue().extract(Format.Builtin.HTTP_HEADERS,
+                        new RequestExtractor(spanHeaders)));
+                if (spanContextMap.get(tracerEntry.getKey()) == null) {
+                    hasParent = false;
+                }
+            }
+
+            if (hasParent) {
+                parentSpanId = UUID.randomUUID().toString();
```

Comment:
New parentId should be generated when there is no parent right? I.e., !hasParent should be condition right? Please check.

Reply:
In the extract method, If we have a parent we create a new Id for it and save it in the spanStore so that child of references can be created to it. If no parent we will mark a ROOT_CONTEXT so that child of relationships are not referenced to this